From 2f82a6bbefb58b3988cd94a99d5dd01fb1452bd0 Mon Sep 17 00:00:00 2001
From: Ivan Vecera <ivecera@redhat.com>
Date: Mon, 8 Sep 2014 16:14:34 -0400
Subject: [ethernet] be2net: Add a dma_mapping_error counter in ethtool

Message-id: <1410192881-30702-72-git-send-email-ivecera@redhat.com>
Patchwork-id: 92448
O-Subject: [RHEL7.1 PATCH 71/78] be2net: Add a dma_mapping_error counter in ethtool
Bugzilla: 1088575
RH-Acked-by: Stefan Assmann <sassmann@redhat.com>
RH-Acked-by: Nikolay Aleksandrov <nikolay@redhat.com>

BZ: #1088575

Upstream commit(s):
commit d3de154071c14549c4b99d06a063bf3761e482c3
Author: Vasundhara Volam <vasundhara.volam@emulex.com>
Date:   Tue Sep 2 09:56:50 2014 +0530

    be2net: Add a dma_mapping_error counter in ethtool

    Add a dma_mapping_error counter to count the number of packets dropped
    due to DMA mapping errors.

    Signed-off-by: Vasundhara Volam <vasundhara.volam@emulex.com>
    Signed-off-by: Sathya Perla <sathya.perla@emulex.com>
    Signed-off-by: David S. Miller <davem@davemloft.net>

Signed-off-by: Ivan Vecera <ivecera@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/emulex/benet/be.h b/drivers/net/ethernet/emulex/benet/be.h
index 4e9bb1f..e5ba83e 100644
--- a/drivers/net/ethernet/emulex/benet/be.h
+++ b/drivers/net/ethernet/emulex/benet/be.h
@@ -323,6 +323,7 @@ struct be_rx_obj {
 struct be_drv_stats {
  u32 be_on_die_temperature;
  u32 eth_red_drops;
+ u32 dma_map_errors;
  u32 rx_drops_no_pbuf;
  u32 rx_drops_no_txpb;
  u32 rx_drops_no_erx_descr;
diff --git a/drivers/net/ethernet/emulex/benet/be_ethtool.c b/drivers/net/ethernet/emulex/benet/be_ethtool.c
index af2ca18..32b432f 100644
--- a/drivers/net/ethernet/emulex/benet/be_ethtool.c
+++ b/drivers/net/ethernet/emulex/benet/be_ethtool.c
@@ -119,6 +119,8 @@ static const struct be_ethtool_stat et_stats[] = {
   * is more than 9018 bytes
   */
  {DRVSTAT_INFO(rx_drops_mtu)},
+ /* Number of dma mapping errors */
+ {DRVSTAT_INFO(dma_map_errors)},
  /* Number of packets dropped due to random early drop function */
  {DRVSTAT_INFO(eth_red_drops)},
  {DRVSTAT_INFO(be_on_die_temperature)},
diff --git a/drivers/net/ethernet/emulex/benet/be_main.c b/drivers/net/ethernet/emulex/benet/be_main.c
index 9657d69..891ad9e 100644
--- a/drivers/net/ethernet/emulex/benet/be_main.c
+++ b/drivers/net/ethernet/emulex/benet/be_main.c
@@ -849,6 +849,7 @@ dma_err:
   unmap_tx_frag(dev, wrb, map_single);
   map_single = false;
   copied -= wrb->frag_len;
+  adapter->drv_stats.dma_map_errors++;
   queue_head_inc(txq);
  }
  return 0;
@@ -1873,7 +1874,7 @@ static void be_post_rx_frags(struct be_rx_obj *rxo, gfp_t gfp)
    if (dma_mapping_error(dev, page_dmaaddr)) {
     put_page(pagep);
     pagep = NULL;
-    rx_stats(rxo)->rx_post_fail++;
+    adapter->drv_stats.dma_map_errors++;
     break;
    }
    page_offset = 0;
-- 
1.7.1